src/mono/mono/metadata/bundled-resources.c

```diff
@@ -0,0 +1,193 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//
+
+#include <glib.h>
+
+#include "bundled-resources-internal.h"
+#include "assembly-internals.h"
+
+#include <mono/metadata/assembly.h>
+#include <mono/metadata/mono-private-unstable.h>
+
+static GHashTable *bundled_resources = NULL;
+
+static int *resources_bundled_counts = NULL;
+
+//---------------------------------------------------------------------------------------
+//
+// mono_free_bundled_resources frees all memory allocated for bundled resources.
+// It should only be called when the runtime no longer needs access to the data,
+// most likely to happen during runtime shutdown.
+//
+
+void
+mono_free_bundled_resources (void)
+{
+	if (bundled_resources)
+		g_hash_table_destroy (bundled_resources);
+	bundled_resources = NULL;
+
+	g_free (resources_bundled_counts);
+	resources_bundled_counts = NULL;
+}
+
+//---------------------------------------------------------------------------------------
+//
+// mono_add_bundled_resource handles bundling of many types of resources to circumvent
+// needing to find or have those resources on disk. The BundledResource struct models
+// the union of information carried by all supported types of resources which are
+// enumerated in MonoBundledResourceType.
+//
+// bundled_resources:
+// A single hash table will hold all resources being bundled with the understanding
+// that all resources being bundled are uniquely named. The bundled resource is tagged
+// with the type of resource it represents for later processing of specific types of
+// resources as done in mono_register_bundled_resources.
+//
+// resources_bundled_counts:
+// An integer array tracks the number of each resource type added to the bundle
+// to facilitate appropriate memory allocation when processing the bundle as done in
+// mono_register_bundled_resources
+//
+// Arguments:
+//  * name - Unique name of the resource
+//  * culture - Culture of satellite assemblies
+//  * data - Byte data array of the bundled resource
+//  * size - Size of byte data array of the bundled resource
+//  type - Type of bundled resource
+//
+
+void
+mono_add_bundled_resource (const char *name, const char *culture, const unsigned char *data, unsigned int size, MonoBundledResourceType type)
+{
```

Comment:
This method should take an array of MonoBundledResource ** and a len and for each item, check type, cast it to the right struct type, get name and add name + ptr to MonoBundledResource to hash map. If type is assembly, add a pointer to the MonoBundledAssembly to the dn_vector_ptr_t for assemblies or if its an MonoBundledSatellite assembly, add pointer to MonoBundledSatelliteAssembly to the dn_vector_ptr_t for satellite assemblies.

We should probably try to handle PDB's in similar way, adding it to the hash map and let embedders use name of pdb file and have another that is the name of the matching assembly in the passed in struct.

Reply:
Modified and folded PDBs/symfiles into the associated assembly/satellite assembly resource